### PR TITLE
Add metrics and tracing test

### PR DESCRIPTION
**Pull request by ericgribkoff**

### CODE REVIEW

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ObservabilityTest {
+
+  private static final int EXPORT_WAIT_SEC = 15;
+
+  private static final Set<String> METRICS =
```
Comment (voidzcy) [inline java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:37]: nit: may want to use lower case instead, or make the set immutable?

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+              "grpc.io/server/sent_messages_per_rpc",
+              "grpc.io/server/received_messages_per_rpc",
+              "grpc.io/server/server_latency"));
+
+  private Server server;
+  private ManagedChannel channel;
+  private RecordingMetricsExporter recordingMetricsExporter;
+  private RecordingTraceExporter recordingTraceExporter;
+
+  private static class WalletImpl extends WalletGrpc.WalletImplBase {
```
Comment (voidzcy) [inline java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:58]: nit: put inner classes in the bottom (although not explicitly specified in Java style guide, most people follows the convention with the _principle of least surprise_ and consider inner classes are less important ([SO link](https://stackoverflow.com/questions/20180533/what-is-the-convention-for-the-placement-of-inner-classes-in-java))).

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    @Override
+    public void fetchBalance(
+        BalanceRequest request, StreamObserver<BalanceResponse> responseObserver) {
+      responseObserver.onNext(BalanceResponse.getDefaultInstance());
+      responseObserver.onCompleted();
+    }
+  }
+
+  private class RecordingTraceExporter extends SpanExporter.Handler {
+    SpanData clientSpanData = null;
```
Comment (voidzcy) [inline java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:68]: nit: we may want some comment maybe? Or some description for traces at the top of this class? It's not something obvious at the first glance (clientSpan is the one for the RPC from wallet client to wallet server and serverSpan is the one for the RPC from wallet server to account server).

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    recordingMetricsExporter.latch.await(EXPORT_WAIT_SEC, TimeUnit.SECONDS);
+    assertThat(recordingMetricsExporter.metrics).containsAtLeastElementsIn(METRICS);
+  }
+
+  @Test
+  public void testTrace() throws Exception {
+    WalletGrpc.newBlockingStub(channel).fetchBalance(BalanceRequest.getDefaultInstance());
+
+    recordingTraceExporter.latch.await(EXPORT_WAIT_SEC, TimeUnit.SECONDS);
+
```
Comment (voidzcy) [inline java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:153]: nit: delete empty line.

**Review by voidzcy — COMMENTED** [review]
Most comments are just suggestive, mostly for stylish things.

Btw, I am feel a little skeptical for the usefulness of such a test. Is it just for providing users some insights about how to use Census metrics and tracing?

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:34]: Fixed

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java (hunk 1 above)
Comment (ericgribkoff) [reply java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:40]: Made immutable

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java (hunk 4 above)
Comment (ericgribkoff) [reply java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java]: done

File: java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java (hunk 3 above)
Comment (ericgribkoff) [reply java/src/test/java/io/grpc/examples/wallet/ObservabilityTest.java:114]: Done

**Review by ericgribkoff — COMMENTED** [review]
Thanks for the review. Addressed the comments.

Regarding the usefulness of the tests: the trace+metrics integration is going into our public TD+gRPC docs, so we are required to have a test verifying this functionality. These do actually serve a purpose beyond just a further how-to example (which these provide as well) - we don't have tests anywhere else that verify the set of metric views that we document are actually the ones sent out (gRPC's tests just verify the raw underlying metrics are sent to OC, and it's up to the user to determine how those should be exported).

**Review by voidzcy — APPROVED** [review]

**Pull request merged**